### PR TITLE
[PowerPC] Mask constant operands in ValueBit tracking

### LINKED ISSUE #59074: [PowerPC] Backend crash due to index out of bound when `lshr` / `shl` i128 vector after `sub`
## Description

When targeting `ppc64`, `ppc32`, or `ppc32le`, the following code containing i128 vector shift crashes due to index being out of bound during `PowerPC DAG->DAG Pattern Instruction Selection`.

The problem occurs for `lshr` and `shl` but not for `ashr`.
The problem occurs for `sub` but not for `add` and the result of `sub` must be used as the shift amount operand.
The problem doesn't occur for i64 vectors.
The problem doesn't occur when targeting `ppc64le`, `aarch64`, `x86_64`, or `riscv64`.

## Minimal Reproduction

https://godbolt.org/z/vW4YT3Ga7

### Code

```ll
define void @f(ptr %0) {
BB:
  %V1 = load <2 x i128>, <2 x i128>* %0
  %V2 = insertelement <2 x i128> %V1, i128 12, i32 0
  %V3 = sub <2 x i128> %V1, %V2
  %V4 = lshr <2 x i128> %V2, %V3
  store <2 x i128> %V4, <2 x i128>* %0
  ret void
}
```

### Stack Trace

```console
llc: /home/henry/aflplusplus-isel/llvm-project/llvm/include/llvm/ADT/SmallVector.h:298: llvm::SmallVectorTemplateCommon::const_reference llvm::SmallVectorTemplateCommon<(anonymous namespace)::BitPermutationSelector::ValueBit>::operator[](llvm::SmallVectorTemplateCommon::size_type) const [T = (anonymous namespace)::BitPermutationSelector::ValueBit]: Assertion `idx < size()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: ./llvm-project/build-debug/bin/llc -mtriple=ppc64 ./crash-reports/dagisel-ppc64/1.ll
1.	Running pass 'Function Pass Manager' on module './crash-reports/dagisel-ppc64/1.ll'.
2.	Running pass 'PowerPC DAG->DAG Pattern Instruction Selection' on function '@f'
 #0 0x00000000047db7da llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:569:11
 #1 0x00000000047db98b PrintStackTraceSignalHandler(void*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:636:1
 #2 0x00000000047d9fc6 llvm::sys::RunSignalHandlers() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Signals.cpp:104:5
 #3 0x00000000047dc0b5 SignalHandler(int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f3802601980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #5 0x00007f38014f1e87 raise /build/glibc-CVJwZb/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #6 0x00007f38014f37f1 abort /build/glibc-CVJwZb/glibc-2.27/stdlib/abort.c:81:0
 #7 0x00007f38014e33fa __assert_fail_base /build/glibc-CVJwZb/glibc-2.27/assert/assert.c:89:0
 #8 0x00007f38014e3472 (/lib/x86_64-linux-gnu/libc.so.6+0x30472)
 #9 0x000000000217ced9 llvm::SmallVectorTemplateCommon<(anonymous namespace)::BitPermutationSelector::ValueBit, void>::operator[](unsigned long) const /home/henry/aflplusplus-isel/llvm-project/llvm/include/llvm/ADT/SmallVector.h:0:5
#10 0x000000000217b0ba (anonymous namespace)::BitPermutationSelector::getValueBits(llvm::SDValue, unsigned int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:1507:21
#11 0x000000000217b469 (anonymous namespace)::BitPermutationSelector::getValueBits(llvm::SDValue, unsigned int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:1543:30
#12 0x0000000002179acc (anonymous namespace)::BitPermutationSelector::Select(llvm::SDNode*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:2674:9
#13 0x0000000002170d66 (anonymous namespace)::PPCDAGToDAGISel::tryBitPermutation(llvm::SDNode*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:3938:17
#14 0x000000000215b184 (anonymous namespace)::PPCDAGToDAGISel::Select(llvm::SDNode*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:5037:7
#15 0x000000000453dac9 llvm::SelectionDAGISel::DoInstructionSelection() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1165:5
#16 0x000000000453cb1a llvm::SelectionDAGISel::CodeGenAndEmitDAG() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:936:3
#17 0x000000000453b4ed llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, bool&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:688:1
#18 0x000000000453af8b llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1603:11
#19 0x0000000004538536 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:467:3
#20 0x000000000215ad79 (anonymous namespace)::PPCDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:167:7
#21 0x0000000003597c35 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:91:8
#22 0x0000000003c7f5e6 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1430:23
#23 0x0000000003c84412 llvm::FPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1476:16
#24 0x0000000003c7feb9 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1545:23
#25 0x0000000003c7fa2d llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:535:16
#26 0x0000000003c846f1 llvm::legacy::PassManager::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1672:3
#27 0x0000000000d2566c compileModule(char**, llvm::LLVMContext&) /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:737:41
#28 0x0000000000d23a12 main /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:418:13
#29 0x00007f38014d4c87 __libc_start_main /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:344:0
#30 0x0000000000d2321a _start (./llvm-project/build-debug/bin/llc+0xd2321a)
```

## Cause

Not sure about the root cause, but the fact the following line returns the shift amount `4294967232` (`-64` represented in `unsigned`) is wrong and caused out-of-bound access to vector in the following code. 

https://github.com/llvm/llvm-project/blob/af029d383a24ee454e1124f64f3427d5b79e5f7e/llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp#L1502

### PR DESCRIPTION
**Pull request by ecnelises**
In IR or C code, shift amount larger than value size is undefined
behavior. But in practice, backend lowering for shift_parts produces
add/sub of shift amounts, thus constant shift amounts might be
negative or larger than value size, which depends on ISA definition.

PowerPC ISA says, the lowest 7 bits (6 bits for 32-bit instruction)
will be taken, and if the highest among them is 1, result will be
zero, otherwise the low 6 bits (or 5 on 32-bit) are used as shift
amount.

This commit emulates the behavior and avoids array overflow in bit
permutation's value bits calculator.

Fixes #59074

### CODE REVIEW

File: llvm/test/CodeGen/PowerPC/pr59074.ll [hunk 1]
```diff
@@ -0,0 +1,132 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=powerpc64le-unknown-linux-gnu -mcpu=pwr7 < %s | FileCheck %s --check-prefix=LE64
+; RUN: llc -mtriple=powerpcle-unknown-linux-gnu -mcpu=pwr7 < %s | FileCheck %s --check-prefix=LE32
+; RUN: llc -mtriple=powerpc64-ibm-aix -mcpu=pwr7 < %s | FileCheck %s --check-prefix=BE64
+; RUN: llc -mtriple=powerpc-ibm-aix -mcpu=pwr7 < %s | FileCheck %s --check-prefix=BE32
```
Comment (ecnelises) [inline llvm/test/CodeGen/PowerPC/pr59074.ll:5]: `-mcpu=pwr8` touches another codepath (such instructions are not generated, so only crash on `-mcpu=pwr7`).

The crash comes from constant shift amounts of `srl_parts`, which only presents after splitting `v2i128` into `v1i128` (from the subtraction parts).

**Comment by jacobly0** [comment]
I was going to report the following bug, but since it is already fixed by this change, maybe this even simpler repro would be useful to add as a test since it doesn't involve any vectors:
```llvm
target triple = "powerpc"
define i64 @entry(ptr) {
  store i6 0, ptr %0
  %2 = load i8, ptr %0
  %3 = and i8 %2, 63
  %4 = zext i8 %3 to i64
  %5 = shl i64 1, %4
  ret i64 %5
}
```
```
$ llc -version
LLVM (http://llvm.org/):
  LLVM version 17.0.6
  DEBUG build with assertions.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: znver4
$ llc repro.ll
llc: llvm/include/llvm/ADT/SmallVector.h:294: reference llvm::SmallVectorTemplateCommon<(anonymous namespace)::BitPermutationSelector::ValueBit>::operator[](size_type) [T = (anonymous namespace)::BitPermutationSelector::ValueBit]: Assertion `idx < size()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: llc repro.ll
1.	Running pass 'Function Pass Manager' on module 'repro.ll'.
2.	Running pass 'PowerPC DAG->DAG Pattern Instruction Selection' on function '@entry'
 #0 0x00007f2d0ca6dd06 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) llvm/lib/Support/Unix/Signals.inc:602:11
 #1 0x00007f2d0ca6e25b PrintStackTraceSignalHandler(void*) llvm/lib/Support/Unix/Signals.inc:675:1
 #2 0x00007f2d0ca6c403 llvm::sys::RunSignalHandlers() llvm/lib/Support/Signals.cpp:104:5
 #3 0x00007f2d0ca6e941 SignalHandler(int) llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007f2d0c2b5c90 (/usr/lib64/libc.so.6+0x39c90)
 #5 0x00007f2d0c30619c (/usr/lib64/libc.so.6+0x8a19c)
 #6 0x00007f2d0c2b5be2 gsignal (/usr/lib64/libc.so.6+0x39be2)
 #7 0x00007f2d0c29e4ed abort (/usr/lib64/libc.so.6+0x224ed)
 #8 0x00007f2d0c29e415 (/usr/lib64/libc.so.6+0x22415)
 #9 0x00007f2d0c2ae542 (/usr/lib64/libc.so.6+0x32542)
#10 0x00007f2d15445499 llvm::SmallVectorTemplateCommon<(anonymous namespace)::BitPermutationSelector::ValueBit, void>::operator[](unsigned long) llvm/include/llvm/ADT/SmallVector.h:0:5
#11 0x00007f2d1544344f (anonymous namespace)::BitPermutationSelector::getValueBits(llvm::SDValue, unsigned int) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:1645:19
#12 0x00007f2d1544395a (anonymous namespace)::BitPermutationSelector::getValueBits(llvm::SDValue, unsigned int) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:1694:30
#13 0x00007f2d15441e59 (anonymous namespace)::BitPermutationSelector::Select(llvm::SDNode*) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:2825:9
#14 0x00007f2d154385ec (anonymous namespace)::PPCDAGToDAGISel::tryBitPermutation(llvm::SDNode*) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:4096:17
#15 0x00007f2d154211c6 (anonymous namespace)::PPCDAGToDAGISel::Select(llvm::SDNode*) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:5195:7
#16 0x00007f2d111cdbba llvm::SelectionDAGISel::DoInstructionSelection() llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1179:5
#17 0x00007f2d111ccbb5 llvm::SelectionDAGISel::CodeGenAndEmitDAG() llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:950:3
#18 0x00007f2d111cb459 llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, bool&) llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:702:1
#19 0x00007f2d111cae34 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1705:11
#20 0x00007f2d111c82c0 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:482:3
#21 0x00007f2d15420d26 (anonymous namespace)::PPCDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:173:7
#22 0x00007f2d1046b4cc llvm::MachineFunctionPass::runOnFunction(llvm::Function&) llvm/lib/CodeGen/MachineFunctionPass.cpp:91:8
#23 0x00007f2d0d427815 llvm::FPPassManager::runOnFunction(llvm::Function&) llvm/lib/IR/LegacyPassManager.cpp:1435:23
#24 0x00007f2d0d42cd4f llvm::FPPassManager::runOnModule(llvm::Module&) llvm/lib/IR/LegacyPassManager.cpp:1481:16
#25 0x00007f2d0d42818b (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) llvm/lib/IR/LegacyPassManager.cpp:1550:23
#26 0x00007f2d0d427c9a llvm::legacy::PassManagerImpl::run(llvm::Module&) llvm/lib/IR/LegacyPassManager.cpp:535:16
#27 0x00007f2d0d42d051 llvm::legacy::PassManager::run(llvm::Module&) llvm/lib/IR/LegacyPassManager.cpp:1677:3
#28 0x000055e89c38576f compileModule(char**, llvm::LLVMContext&) llvm/tools/llc/llc.cpp:754:41
#29 0x000055e89c3839d5 main llvm/tools/llc/llc.cpp:416:13
#30 0x00007f2d0c29feea (/usr/lib64/libc.so.6+0x23eea)
#31 0x00007f2d0c29ffa5 __libc_start_main (/usr/lib64/libc.so.6+0x23fa5)
#32 0x000055e89c3831a1 _start (bin/llc+0x2e1a1)
Aborted (core dumped)
```

File: llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp [hunk 2]
```diff
@@ -1635,7 +1635,7 @@ class BitPermutationSelector {
     default: break;
     case ISD::ROTL:
       if (isa<ConstantSDNode>(V.getOperand(1))) {
-        unsigned RotAmt = V.getConstantOperandVal(1);
+        unsigned RotAmt = V.getConstantOperandVal(1) & (NumBits - 1);
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:1638]:
```suggestion
        assert(isPowerOf2_32(NumBits) && "...");
        unsigned RotAmt = V.getConstantOperandVal(1) & (NumBits - 1);
```

**Review by bzEq — APPROVED** [review]
LGTM with nit.

**Comment by Rexicon226** [comment]
Thank you @bzEq for the quick response. Would it be possible to backport this into 18?

**Comment by ecnelises** [comment]
> Thank you @bzEq for the quick response. Would it be possible to backport this into 18?

/cherry-pick 292d9e8

**Pull request merged**